tools/src/main/scala/scala/scalanative/codegen/Lower.scala

```diff
@@ -154,18 +170,33 @@ object Lower {
 
       buf ++= handlers
 
-      eliminateDeadCode(buf.toSeq.map(super.onInst))
+      eliminateDeadCode {
+        buf.toSeq.map(super.onInst)
+      }
     }
 
     override def onVal(value: Val): Val = value match {
-      case Val.Global(ScopeRef(node), _) =>
-        Val.Global(rtti(node).name, Type.Ptr)
-      case Val.String(v) =>
-        genStringVal(v)
-      case Val.Unit =>
-        unit
-      case _ =>
-        super.onVal(value)
+      case Val.ClassOf(_) =>
+        util.unsupported("Lowering ClassOf needs nir.Buffer")
+      case Val.Global(ScopeRef(node), _) => rtti(node).const
+      case Val.String(v)                 => genStringVal(v)
+      case Val.Unit                      => unit
+      case _                             => super.onVal(value)
+    }
+
+    def genClassOf(buf: Buffer, node: ScopeInfo): Val = {
+      val tpePtr = Val.Const(rtti(node).const)
+      buf.call(JavaGetClassSig, JavaGetClass, Seq(tpePtr), Next.None)
```

Comment:
This does not seem right. It seems you're calling `java.lang.Object.getClass` to get a `j.l.Class[_]`. But you don't have any object to use as the `this` object. IIUC this line passes `tpePtr` itself as the receiver, but that would not be type-correct, since `tpePtr` is a raw pointer to a C struct, not an instance of `java.lang.Object` or its subclasses.

This should be calling `runtime.toClass` (which I suggested to create above) with the singleton instance of `runtime.package$` as receiver, and `typePtr` as argument.